operator/autoscaler.go

```diff
@@ -225,46 +225,59 @@ func (as *AutoScaler) getMaxDiskUsage(managedNodes []ESNode) float64 {
 	return maxDisk
 }
 
-func (as *AutoScaler) ensureLowerBoundNodeReplicas(scalingSpec *zv1.ElasticsearchDataSetScaling, newDesiredNodeReplicas int32) int32 {
-	if scalingSpec.MinReplicas > 0 {
-		return int32(math.Max(float64(newDesiredNodeReplicas), float64(scalingSpec.MinReplicas)))
-	}
-	return newDesiredNodeReplicas
-}
-
-func (as *AutoScaler) ensureUpperBoundNodeReplicas(scalingSpec *zv1.ElasticsearchDataSetScaling, newDesiredNodeReplicas int32) int32 {
+func (as *AutoScaler) ensureBoundsNodeReplicas(scalingSpec *zv1.ElasticsearchDataSetScaling, newDesiredNodeReplicas int32) int32 {
 	if scalingSpec.MaxReplicas > 0 {
-		upperBound := int32(math.Min(float64(newDesiredNodeReplicas), float64(scalingSpec.MaxReplicas)))
-		if upperBound < newDesiredNodeReplicas {
+		newDesiredNodeReplicas = int32(math.Min(float64(newDesiredNodeReplicas), float64(scalingSpec.MaxReplicas)))
+		if int32(scalingSpec.MaxReplicas) < newDesiredNodeReplicas {
 			as.logger.Warnf("Requested to scale up to %d, which is beyond the defined maxReplicas of %d.", newDesiredNodeReplicas, scalingSpec.MaxReplicas)
+			return newDesiredNodeReplicas
```

Comment:
This looks wrong to me. Are we going to allow scaling beyond scalingSpec.MaxReplicas?

Reply:
I think it's just a mistake and the total logic in this function is not correct
```
		newDesiredNodeReplicas = int32(math.Min(float64(newDesiredNodeReplicas), float64(scalingSpec.MaxReplicas)))
		if int32(scalingSpec.MaxReplicas) < newDesiredNodeReplicas {
			as.logger.Warnf("Requested to scale up to %d, which is beyond the defined maxReplicas of %d.", newDesiredNodeReplicas, scalingSpec.MaxReplicas)
			return newDesiredNodeReplicas
		}
```
I will update it

Reply:
I updated this function and add a test for it
https://github.com/zalando-incubator/es-operator/pull/157/files#diff-add97f45040e000fdf77c6cab0697b142d788e56b1379176785662a659a5186bR228